---
title: "P337: Compare module expression in BAL and BE samples"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---

# Background

The purpose of this workflow is to define the epithelial and cellular expression patterns that relate to eosinophil recruitment into the airway after allergen challenge. 

These data were previously cleaned and modules were built from genes significant for allergen challenges. Briefly, low quality libraries were removed (median CV > 0.65, alignment < 80%, sequences < 1E6). Genes were filtered to protein-coding expressed at > 1 CPM in at least 10% of BAL and BE samples separately. TMM and voom log2 normalization was then completed. Modules were built from genes that changed significantly with allergen challenge (FDR < 0.3) in BE and for allergen challenge and cell type (FDR < 0.3) in BAL samples. Epithelial cell associated modules were removed from BAL analysis since most BAL samples had 0% epithelial cells. For details, see:

* BAL: <https://github.com/altman-lab/P337_MINA_BAL_public> 
* BE: <https://github.com/altman-lab/P337_MINA_BE_public>

# Setup

Load packages

```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
# RNAseq data
library(limma)
library(kimma)
# Heatmap
library(ComplexHeatmap)
library(circlize)
# Correlation
library(Hmisc)
library(corrplot)
# Print pretty table to knit file
library(knitr)
library(kableExtra)
options(knitr.kable.NA = '')

`%notin%` <- Negate(`%in%`)
```

Set seed

```{r}
set.seed(4389)
```

# Data
## Load data

```{r}
# Genes
attach("data_clean/P337_BAL_data.RData")
dat.BAL <- dat.BAL.abund.norm.voom
attach("data_clean/P337_BE_data.RData")
dat.BE <- dat.BE.abund.norm.voom

# Modules
# mod.genes.BAL <- read_csv("data_clean/P337_ALL_genes_in_mod.csv")
mod.voom.BAL <- read_csv("data_clean/P337_ALL_mod_voom_counts.csv") %>% 
  mutate(module = gsub("module_P337_","BAL_",module)) 
# mod.genes.BE <- read_csv("data_clean/P337_BE_genes_in_mod.csv")
mod.voom.BE <- read_csv("data_clean/P337_BE_mod_voom_counts.csv") %>% 
  mutate(module = gsub("module_P337_","",module))
```

## Calculate V5 - V4

When applicable, correlations will be completed for delta values (V5-V4) to capture the paired nature of these data.

#### Gene expression

```{r}
dat.BAL.delta <- as.data.frame(dat.BAL$E) %>% 
  rownames_to_column("geneName") %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  left_join(dat.BAL$targets %>% 
              select(libID, donorID, visit)) %>% 
  select(-libID) %>% 
  pivot_wider(names_from = visit) %>% 
  mutate(delta = V5-V4) %>% 
  select(-V4,-V5) %>% 
  drop_na(delta) %>% 
  pivot_wider(names_from = donorID, values_from = delta)
  
dat.BE.delta <- as.data.frame(dat.BE$E) %>% 
  rownames_to_column("geneName") %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  left_join(dat.BE$targets %>% 
              select(libID, donorID, visit)) %>% 
  select(-libID) %>% 
  pivot_wider(names_from = visit) %>% 
  mutate(delta = V5-V4) %>% 
  select(-V4,-V5) %>% 
  drop_na(delta) %>% 
  pivot_wider(names_from = donorID, values_from = delta)
```

#### Module expression

```{r}
mod.BAL.delta <- mod.voom.BAL %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BAL$targets %>% 
              select(libID, donorID, visit)) %>% 
  select(-libID) %>% 
  pivot_wider(names_from = visit) %>% 
  mutate(delta = V5-V4) %>% 
  select(-V4,-V5) %>% 
  drop_na(delta) %>% 
  pivot_wider(names_from = donorID, values_from = delta) %>% 
  column_to_rownames("module")
  
mod.BE.delta <- mod.voom.BE %>% 
  select(-module.char) %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BE$targets %>% 
              select(libID, donorID, visit)) %>% 
  select(-libID) %>% 
  pivot_wider(names_from = visit) %>% 
  mutate(delta = V5-V4) %>% 
  select(-V4,-V5) %>% 
  drop_na(delta) %>% 
  pivot_wider(names_from = donorID, values_from = delta) %>% 
  column_to_rownames("module")
```

# Compare BAL and BE modules
## Cluster by module expression

Rename module expression by donor.

```{r}
mod.BAL.rename <- mod.voom.BAL %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BAL$targets %>% 
              select(libID, donorID, visit)) %>% 
  mutate(name = paste(donorID, visit, sep = "_")) %>% 
  select(-c(libID, donorID, visit))

mod.BE.rename <- mod.voom.BE %>% 
  select(-module.char) %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BE$targets %>% 
              select(libID, donorID, visit)) %>% 
  mutate(name = paste(donorID, visit, sep = "_")) %>% 
  select(-c(libID, donorID, visit))

#Combine and remove donors not in both datasets
lib.overlap <- intersect(mod.BAL.rename$name,mod.BE.rename$name)

mod.all.rename <- bind_rows(mod.BAL.rename, mod.BE.rename) %>% 
  filter(name %in% lib.overlap) %>% 
  arrange(name) %>% 
  pivot_wider() %>% 
  column_to_rownames("module")
```

```{r fig.height=8}
visit.anno <- strsplit(colnames(mod.all.rename), split = "_") %>% unlist()
visit.anno <- visit.anno[seq(2, length(visit.anno), by = 2)]
hm.anno <- HeatmapAnnotation(
    visit = visit.anno, 
    col = list(visit = c("V4" = "#E1BE6A", "V5" = "#40B0A6")
    ))

Heatmap(as.matrix(mod.all.rename), top_annotation = hm.anno,
        row_km = 4, column_km = 2, name = "log2 CPM")

Heatmap(scale(mod.all.rename), top_annotation = hm.anno,
        row_km = 4, column_km = 2, name = "Z-score")
```

## Pearson correlation of delta values
Filter donors with both sample types.

```{r}
donor.overlap <- intersect(colnames(mod.BAL.delta), colnames(mod.BE.delta))

mod.BAL.delta.overlap <- mod.BAL.delta %>% 
  select(all_of(donor.overlap))

mod.BE.delta.overlap <- mod.BE.delta %>% 
  select(all_of(donor.overlap))
```

```{r}
corr.delta <- rcorr(t(mod.BAL.delta.overlap), t(mod.BE.delta.overlap),
                     type = "pearson")

corr.delta.R <- as.data.frame(corr.delta$r) %>% 
  rownames_to_column() %>% 
  filter(grepl("BAL", rowname)) %>% 
  column_to_rownames() %>% 
  select(contains("BE")) %>% 
  as.matrix()
corr.delta.P <- as.data.frame(corr.delta$P) %>% 
  rownames_to_column() %>% 
  filter(grepl("BAL", rowname)) %>% 
  column_to_rownames() %>% 
  select(contains("BE")) %>% 
  as.matrix()

#heatmap
corrplot(corr.delta.R, p.mat = corr.delta.P, is.corr = FALSE,
         method = 'color',
         sig.level = c(0.001, 0.01, 0.05), pch.cex = 0.9,
         insig = 'label_sig', pch.col = 'grey20',
         tl.col = "black", col = rev(COL2('RdBu')), col.lim = c(-1,1))

```

# Linear modeling

## EOS percentage interaction

```{r}
mod.voom.BAL.mat <- mod.voom.BAL %>% 
  column_to_rownames("module") %>% 
  select(all_of(dat.BAL$targets$libID))
  
fdr.EOSint.BAL <- kmFit(counts = mod.voom.BAL.mat, meta = dat.BAL$targets,
                     patientID = "donorID",
                     model = "~visit*EOS.pct + (1|donorID)",
                     run.lme = TRUE, run.contrast = TRUE, contrast.var = "visit:EOS.pct")
```

```{r}
mod.voom.BE.mat <- mod.voom.BE %>% 
  column_to_rownames("module") %>% 
  select(all_of(dat.BE$targets$libID))

fdr.EOSint.BE <- kmFit(counts = mod.voom.BE.mat, meta = dat.BE$targets,
                     patientID = "donorID",
                     model = "~visit*EOS.pct + (1|donorID)",
                     run.lme = TRUE, run.contrast = TRUE, contrast.var = "visit:EOS.pct")
```

**BAL significant modules**

```{r}
summarise_kmFit(fdr.EOSint.BAL$lme)
summarise_kmFit(fdr.EOSint.BAL$lme.contrast)
```

**BE significant modules**

```{r}
summarise_kmFit(fdr.EOSint.BE$lme)
summarise_kmFit(fdr.EOSint.BE$lme.contrast)
```

### Visualize significant modules

```{r}
BAL.sig <- fdr.EOSint.BAL$lme %>% 
  filter(variable == "visit:EOS.pct" & FDR < 0.05) %>% 
  pull(gene)

BE.sig <- fdr.EOSint.BE$lme %>% 
  filter(variable == "visit:EOS.pct" & FDR < 0.05) %>% 
  pull(gene)

#combine mod expression data
mod.voom.BAL.rename <- mod.voom.BAL.mat %>% 
  rownames_to_column("module") %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BAL$targets %>% 
              select(libID, group, donorID, visit, EOS.pct)) %>% 
  mutate(name = paste(group, donorID, visit, sep = "_"))

mod.voom.BE.rename <- mod.voom.BE.mat %>% 
  rownames_to_column("module") %>% 
  pivot_longer(-module, names_to = "libID") %>% 
  left_join(dat.BE$targets %>% 
              select(libID, group, donorID, visit, EOS.pct)) %>% 
  mutate(name = paste(group, donorID, visit, sep = "_"))

bind_rows(mod.voom.BAL.rename, mod.voom.BE.rename) %>% 
  filter(module %in% c(BAL.sig, BE.sig)) %>% 
  
  ggplot() +
  aes(x = EOS.pct, y = value, color = visit) +
  geom_point() +
  geom_smooth(se = FALSE, method = "lm", formula = 'y ~ x') +
  facet_wrap(~ module, scales = "free") +
  labs(x = "EOS percent", y = "log2 CPM") +
  theme_classic()
```

## EOS percentage post-only

```{r}
fdr.EOSint.BAL.post <- kmFit(counts = mod.voom.BAL.mat, meta = dat.BAL$targets,
                     patientID = "donorID",
                     model = "~EOS.pct",
                     run.lm = TRUE, 
                     subset.var = "visit", subset.lvl = "V5")
```

```{r}
fdr.EOSint.BE.post <- kmFit(counts = mod.voom.BE.mat, meta = dat.BE$targets,
                     patientID = "donorID",
                     model = "~EOS.pct",
                     run.lm = TRUE, 
                     subset.var = "visit", subset.lvl = "V5")
```

**BAL significant modules**

```{r}
summarise_kmFit(fdr.EOSint.BAL.post$lm)
```

**BE significant modules**

```{r}
summarise_kmFit(fdr.EOSint.BE.post$lm)
```

### Visualize significant modules

```{r}
BAL.sig2 <- fdr.EOSint.BAL.post$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.05) %>% 
  pull(gene)

BE.sig2 <- fdr.EOSint.BE.post$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.05) %>% 
  pull(gene)

bind_rows(mod.voom.BAL.rename, mod.voom.BE.rename) %>% 
  filter(module %in% c(BAL.sig2, BE.sig2)) %>% 
  filter(visit == "V5") %>% 
  
  ggplot() +
  aes(x = EOS.pct, y = value, color = visit) +
  geom_point() +
  geom_smooth(se = FALSE, method = "lm", formula = 'y ~ x') +
  facet_wrap(~ module, scales = "free") +
  labs(x = "EOS percent", y = "log2 CPM") +
  theme_classic()
```

## FeNO interaction

Reformt FeNO pre and post into 1 column

```{r}
feno.BAL <- dat.BAL$targets %>% 
  select(donorID, FeNO.PreBro_V4, FeNO.PreBro_V5) %>% 
  pivot_longer(-donorID) %>% 
  separate(name, into = c("name", "visit"), sep = "_") %>% 
  distinct() %>% 
  pivot_wider()

dat.BAL$targets <- dat.BAL$targets %>% 
  inner_join(feno.BAL)
```

```{r}
fdr.FeNOint.BAL <- kmFit(counts = mod.voom.BAL.mat, meta = dat.BAL$targets,
                     patientID = "donorID",
                     model = "~visit*FeNO.PreBro + (1|donorID)",
                     run.lme = TRUE, run.contrast = TRUE, contrast.var = "visit:FeNO.PreBro")
```


```{r}
feno.BE <- dat.BE$targets %>% 
  select(donorID, FeNO.PreBro_V4, FeNO.PreBro_V5) %>% 
  pivot_longer(-donorID) %>% 
  separate(name, into = c("name", "visit"), sep = "_") %>% 
  distinct() %>% 
  pivot_wider()

dat.BE$targets <- dat.BE$targets %>% 
  inner_join(feno.BE)
```

```{r}
fdr.FeNOint.BE <- kmFit(counts = mod.voom.BE.mat, meta = dat.BE$targets,
                     patientID = "donorID",
                     model = "~visit*FeNO.PreBro + (1|donorID)",
                     run.lme = TRUE, run.contrast = TRUE, contrast.var = "visit:FeNO.PreBro")
```

**BAL significant modules**

```{r}
summarise_kmFit(fdr.FeNOint.BAL$lme)
summarise_kmFit(fdr.FeNOint.BAL$lme.contrast)
```

**BE significant modules**

```{r}
summarise_kmFit(fdr.FeNOint.BE$lme)
summarise_kmFit(fdr.FeNOint.BE$lme.contrast)
```

### Visualize significant modules

```{r}
BAL.sig3 <- fdr.FeNOint.BAL$lme %>% 
  filter(variable == "visit:FeNO.PreBro" & FDR < 0.05) %>% 
  pull(gene)

BE.sig3 <- fdr.FeNOint.BE$lme %>% 
  filter(variable == "visit:FeNO.PreBro" & FDR < 0.05) %>% 
  pull(gene)

#combine mod expression data
bind_rows(mod.voom.BAL.rename, mod.voom.BE.rename) %>% 
  filter(module %in% c(BAL.sig3, BE.sig3)) %>% 
  left_join(feno.BE) %>% 
  
  ggplot() +
  aes(x = FeNO.PreBro, y = value, color = visit) +
  geom_point() +
  geom_smooth(se = FALSE, method = "lm", formula = 'y ~ x') +
  facet_wrap(~ module, scales = "free") +
  labs(x = "FeNO (PreBro)", y = "log2 CPM") +
  theme_classic()
```

# Save results

```{r}
save(fdr.EOSint.BAL, fdr.EOSint.BAL.post,
     fdr.EOSint.BE, fdr.EOSint.BE.post,
     fdr.FeNOint.BAL, fdr.FeNOint.BE,
     file = "results/module_model_fitting.RData")
```

# R session

```{r}
sessionInfo()
```

***