---
title: "P337: BAL and BE gene expression in response to allergen challenge"
subtitle: "Relationships to eosinophils and FeNO"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE,
                      fig.width = 8.5)
```

# Background

The purpose of this workflow is to define the epithelial and cellular expression patterns that relate to eosinophil recruitment into the airway after allergen challenge. 

These data were previously cleaned and modules were built from genes significant for allergen challenges. Briefly, low quality libraries were removed (median CV > 0.65, alignment < 80%, sequences < 1E6). Genes were filtered to protein-coding expressed at > 1 CPM in at least 10% of BAL and BE samples separately. TMM and voom log2 normalization was then completed. For details, see:

* BAL: <https://github.com/altman-lab/P337_MINA_BAL_public> 
* BE: <https://github.com/altman-lab/P337_MINA_BE_public>


```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
# # RNAseq data
library(limma)
library(kimma)
# #plots
library(BIGpicture)
library(patchwork)
library(ggpubr)
# # Heatmap
library(ComplexHeatmap)
# library(circlize)
# # Correlation
library(Hmisc)
# library(corrplot)
# Enrichment plots
library(rrvgo)
library(GO.db)
# Print pretty table to knit file
library(knitr)
library(kableExtra)
options(knitr.kable.NA = '')

# `%notin%` <- Negate(`%in%`)
select <- dplyr::select
rename <- dplyr::rename
mutate <- dplyr::mutate
```

```{r}
#load data
attach("data_clean/P337_BAL_data.RData")
dat.BAL <- dat.BAL.abund.norm.voom
attach("data_clean/P337_BE_data.RData")
dat.BE <- dat.BE.abund.norm.voom
#load results
load("results/gene_model_fitting.RData")
load("results/DEG_enrichment.RData")
#Combine gene keys
gene.key <- dat.BAL$genes %>% 
  bind_rows(dat.BE$genes) %>% 
  select(geneName, hgnc_symbol) %>% 
  distinct()
# Recode visit
dat.BAL$targets <- dat.BAL$targets %>% 
  mutate(visit = case_when(visit == "V4" ~ "Pre",
                           visit == "V5" ~ "Post"),
         visit = factor(visit, levels = c("Pre","Post")))

dat.BE$targets <- dat.BE$targets %>% 
  mutate(visit = case_when(visit == "V4" ~ "Pre",
                           visit == "V5" ~ "Post"),
         visit = factor(visit, levels = c("Pre","Post")))
```

# Differentially expressed genes (DEG)
## Genes significant for allergen challenge (Post - Pre)

```{r echo=FALSE}
summarise_lmFit(lm_visit_BAL$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
  mutate(sample = "BAL", .before = variable) %>% 
  bind_rows(summarise_lmFit(lm_visit_BE$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
              mutate(sample = "BE", .before = variable)) %>% 
  rename_all(~gsub("_", " < ", .)) %>% 
  filter(variable == "visit") %>% 
  kable(caption = "Post vs Pre allergen challenge") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r}
plot_venn_genes(model_result = list("BAL "=lm_visit_BAL$lm,
                                    "BE "=lm_visit_BE$lm),
                variables = c("visit"), fdr.cutoff = c(0.05,0.1,0.3))$venn %>% 
  wrap_plots()
```

## Genes significant for allergen challenge and eosinophils

Data were filtered to genes significant for allergen challenge at FDR < 0.3. Samples were filtered to only Post allergen challenge as most Pre samples contained 0% eosinophils.

```{r echo=FALSE}
summarise_lmFit(lm_eos_BAL$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
  mutate(sample = "BAL", .before = variable) %>% 
  bind_rows(summarise_lmFit(lm_eos_BE$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
              mutate(sample = "BE", .before = variable)) %>% 
  rename_all(~gsub("_", " < ", .)) %>% 
  filter(variable == "EOS.pct") %>% 
  kable(caption = "Eosinophil percent post allergen challenge") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r}
plot_venn_genes(model_result = list("BAL "=lm_eos_BAL$lm,
                                    "BE "=lm_eos_BE$lm),
                variables = c("EOS.pct"), fdr.cutoff = c(0.05,0.1,0.3))$venn %>% 
  wrap_plots()
```

## Genes significant for allergen challenge and FeNO

As with eosinophil models, data were filtered to genes significant for allergen challenge (FDR < 0.3) and Post allergen challenge samples.

```{r echo=FALSE}
summarise_lmFit(lm_feno_BAL$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
  mutate(sample = "BAL", .before = variable) %>% 
  bind_rows(summarise_lmFit(lm_feno_BE$lm, fdr.cutoff = c(0.05,0.1,0.3), FCgroup = TRUE) %>% 
              mutate(sample = "BE", .before = variable)) %>% 
  rename_all(~gsub("_", " < ", .)) %>% 
  filter(variable == "FeNO") %>% 
  kable(caption = "FeNO post allergen challenge") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r}
plot_venn_genes(model_result = list("BAL"=lm_feno_BAL$lm,
                                    "BE"=lm_feno_BE$lm),
                variables = c("FeNO"), fdr.cutoff = c(0.05,0.1,0.3))$venn %>% 
  wrap_plots()
```

### DEG results files

See `DEG_lists.csv` for genes significant at FDR < 0.05, 0.1, and 0.3. Column names are `sample_variable_FDRcutoff`. Full model results can be found in `gene_lm_results.csv`

## DEG enrichment
### Summary

```{r include=FALSE}
load("results/DEG_enrichment.RData")
enrich.all <- bind_rows(deg_H, deg_C2, deg_C5) %>% 
  separate(group, into = c("Sample","Variable","DEG FDR"), sep = "_",
          remove = FALSE) %>% 
  mutate(gs_cat = factor(gs_cat, levels=c("H","C2","C5")),
         Variable = factor(Variable, levels = c("visit", "eos", "feno")))
```

```{r}
temp <- enrich.all %>% 
  filter(FDR < 0.1) %>% 
  count(group, Sample, Variable, `DEG FDR`, size_group, gs_cat) %>% 
  rename("FDR < 0.1"=n)

enrich.all %>% 
  filter(FDR < 0.01) %>% 
  count(group, Sample, Variable, `DEG FDR`, size_group, gs_cat)  %>%  
  rename("FDR < 0.01"=n) %>% 
  full_join(temp) %>% 
  filter(`DEG FDR`==0.1 | Variable == "visit") %>% 
  arrange(Variable, Sample, `DEG FDR`, gs_cat) %>% 
  select(-group, -Variable, -`DEG FDR`) %>% 
  kable(caption = "Total enriched pathways") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1) %>% 
  pack_rows("Visit DEG, FDR < 0.3", 1,6) %>% 
  pack_rows("EOS DEG, FDR < 0.1", 7,11) %>% 
  pack_rows("FeNO DEG, FDR < 0.1", 12,14)
```

### Eosinophils

Allergen challenge (FDR < 0.3) and eosinophil significant (FDR 0.1) gene sets. The following table includes up to the top 11 most significant (lowest FDR) gene sets in BAL and three databases (Hallmark, C2 canonical pathways, C5 GO biological processes). This includes all BE significant gene sets as well as all BAL H significant gene sets. BAL C2 and C5 lists are reduced by the top 11 cutoff.

```{r}
enrich.all %>% 
  filter(Variable == "eos" & `DEG FDR` == 0.1 & FDR < 0.1) %>% 
  group_by(group, gs_cat) %>% 
  slice_min(order_by = FDR, n = 11, with_ties = TRUE) %>% 
  ungroup() %>% 
  select(Sample, gs_cat, pathway, FDR) %>% 
  mutate(FDR = signif(FDR, digits = 2)) %>% 
  arrange(Sample, gs_cat, FDR) %>% 
  
  kable(caption = "Eosinophil enriched pathways") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:2)
```

C5 GO terms in BAL are further explored based on GO hierarchy since there are so many. Here, we use the gene sets at FDR < 0.01 and group similar terms under their parent terms (similarity > 0.7).

```{r}
c5.db <- as.data.frame(GOTERM)[,c(1,3)] %>% 
  distinct() %>% 
  mutate(Term = gsub("-|,|/", " ", Term),
         Term = gsub("  "," ", Term),
         Term = tolower(Term))

eos.c5.bal <- enrich.all %>% 
  filter(Variable == "eos" & Sample == "BAL" & `DEG FDR` == 0.1 &
           FDR < 0.1 & FDR < 0.01) %>% 
  filter(gs_cat == "C5") %>% 
  select(group, pathway, FDR) %>% 
  mutate(Term = gsub("GOBP_", "", pathway),
         Term = gsub("_", " ", Term),
         Term = tolower(Term)) %>% 
  left_join(c5.db)
```

```{r fig.height=10}
scores <- eos.c5.bal$FDR
  names(scores) <- eos.c5.bal$go_id

simMatrix <- calculateSimMatrix(eos.c5.bal$go_id,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

reducedTerms <- reduceSimMatrix(simMatrix,
                                scores,
                                threshold=0.7,
                                orgdb="org.Hs.eg.db")
treemapPlot(reducedTerms)
```

### FeNO

Allergen challenge (FDR < 0.3) and FeNO significant (FDR 0.1) gene sets. The following table includes all significant gene sets in BAL and BE.

```{r}
enrich.all %>% 
  filter(Variable == "feno" & `DEG FDR` == 0.1 & FDR < 0.1) %>% 
  select(Sample, gs_cat, pathway, FDR) %>% 
  mutate(FDR = signif(FDR, digits = 2)) %>% 
  arrange(Sample, gs_cat, FDR) %>% 
  
  kable(caption = "FeNO enriched pathways") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:2)
```

# Comparing gene expression in BAL and BE
## Eosinophils

Allergen challenge (FDR < 0.3) and eosinophil significant (FDR 0.1) genes in BAL vs BE. As expected, genes can be grouped into those that positively associate with eosinophils in both BAL and BE (positive correlation) and those that have opposite directionality in BAL and BE (negative correlation).

```{r}
dat.BAL.rename <- as.data.frame(dat.BAL$E) %>% 
  rownames_to_column("gene") %>% 
  pivot_longer(-gene, names_to = "libID") %>% 
  left_join(dat.BAL$targets %>% 
              select(libID, donorID, visit)) %>% 
  filter(visit == "Post") %>% 
  select(gene, donorID, value) %>% 
  mutate(gene = paste("BAL",gene,sep="_"))

dat.BE.rename <- as.data.frame(dat.BE$E) %>% 
  rownames_to_column("gene") %>% 
  pivot_longer(-gene, names_to = "libID") %>% 
  left_join(dat.BE$targets %>% 
              select(libID, donorID, visit)) %>% 
  filter(visit == "Post") %>% 
  select(gene, donorID, value) %>% 
  mutate(gene = paste("BE",gene,sep="_"))
```

```{r}
#EOS signif genes
BAL_eos_0.1_up <- lm_eos_BAL$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.1 & estimate > 0) %>% 
  mutate(gene = paste("BAL",gene,sep="_")) %>% 
  pull(gene)
BAL_eos_0.1_dn <- lm_eos_BAL$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.1 & estimate < 0) %>% 
  mutate(gene = paste("BAL",gene,sep="_")) %>% 
  pull(gene)
BE_eos_0.1_up <- lm_eos_BE$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.1 & estimate > 0) %>% 
  mutate(gene = paste("BE",gene,sep="_")) %>% 
  pull(gene)
BE_eos_0.1_dn <- lm_eos_BE$lm %>% 
  filter(variable == "EOS.pct" & FDR < 0.1 & estimate < 0) %>% 
  mutate(gene = paste("BE",gene,sep="_")) %>% 
  pull(gene)

#FeNO signif genes
BAL_feno_0.1_up <- lm_feno_BAL$lm %>% 
  filter(variable == "FeNO" & FDR < 0.1 & estimate > 0) %>% 
  mutate(gene = paste("BAL",gene,sep="_")) %>% 
  pull(gene)
BAL_feno_0.1_dn <- lm_feno_BAL$lm %>% 
  filter(variable == "FeNO" & FDR < 0.1 & estimate < 0) %>% 
  mutate(gene = paste("BAL",gene,sep="_")) %>% 
  pull(gene)
BE_feno_0.1_up <- lm_feno_BE$lm %>% 
  filter(variable == "FeNO" & FDR < 0.1 & estimate > 0) %>% 
  mutate(gene = paste("BE",gene,sep="_")) %>% 
  pull(gene)
BE_feno_0.1_dn <- lm_feno_BE$lm %>% 
  filter(variable == "FeNO" & FDR < 0.1 & estimate < 0) %>% 
  mutate(gene = paste("BE",gene,sep="_")) %>% 
  pull(gene)

save(BAL_eos_0.1_up,BAL_eos_0.1_dn,BE_eos_0.1_up,BE_eos_0.1_dn,
     BAL_feno_0.1_up,BAL_feno_0.1_dn,BE_feno_0.1_up,BE_feno_0.1_dn,
     file="results/signif_gene_lists.RData")
```

```{r}
#Combine 
temp <- dat.BAL.rename %>% 
  filter(gene %in% c(BAL_eos_0.1_up, BAL_eos_0.1_dn)) %>% 
  mutate(EOS = ifelse(gene %in% c(BAL_eos_0.1_up, BAL_eos_0.1_dn), "FDR < 0.1", "NS"),
         FeNO = ifelse(gene %in% c(BAL_feno_0.1_up, BAL_feno_0.1_dn), "FDR < 0.1", "NS"))

dat.visit.all.rename <- dat.BE.rename %>% 
  filter(gene %in% c(BE_eos_0.1_up, BE_eos_0.1_dn)) %>% 
  mutate(EOS = ifelse(gene %in% c(BE_eos_0.1_up, BE_eos_0.1_dn), "FDR < 0.1", "NS"),
         FeNO = ifelse(gene %in% c(BE_feno_0.1_up, BE_feno_0.1_dn), "FDR < 0.1", "NS")) %>% 
  bind_rows(temp) %>% 
  arrange(donorID) %>% 
  pivot_wider(names_from = donorID) 
```

```{r}
hm.mat.up <- dat.visit.all.rename %>% 
  select_if(~all(!is.na(.))) %>%
  filter(gene %in% c(BAL_eos_0.1_up,BE_eos_0.1_up))

hm.mat.up.mat <- hm.mat.up %>% 
  select(-EOS, -FeNO) %>% 
  column_to_rownames("gene")

hm.mat.dn <- dat.visit.all.rename %>% 
  select_if(~all(!is.na(.))) %>%
  filter(gene %in% c(BAL_eos_0.1_dn,BE_eos_0.1_dn)) 

hm.mat.dn.mat <- hm.mat.dn %>% 
  select(-EOS, -FeNO) %>% 
  column_to_rownames("gene")
```

```{r}
corr.up <- rcorr(t(hm.mat.up.mat), type = "pearson")
corr.dn <- rcorr(t(hm.mat.dn.mat), type = "pearson")

corr.up.R <- as.data.frame(corr.up$r) %>% 
  rownames_to_column() %>% 
  pivot_longer(-rowname) %>% 
  #filter BAL vs BE
  filter(grepl("BAL_", rowname)) %>% 
  filter(grepl("BE_", name)) %>% 
  mutate(across(c(rowname, name), ~gsub("BAL_|BE_", "", .))) %>% 
  #Rename columns
  left_join(gene.key, by=c("name"="geneName")) %>% 
  mutate(name = paste0(hgnc_symbol, " (", name, ")")) %>% 
  select(-hgnc_symbol) %>% 
  pivot_wider() %>% 
  #rename rows
  left_join(gene.key, by=c("rowname"="geneName")) %>% 
  mutate(rowname = paste0(hgnc_symbol, " (", rowname, ")")) %>% 
  select(-hgnc_symbol) %>% 
  column_to_rownames() %>% 
  as.matrix()

corr.dn.R <- as.data.frame(corr.dn$r) %>% 
  rownames_to_column() %>% 
  pivot_longer(-rowname) %>% 
  #filter BAL vs BE
  filter(grepl("BAL_", rowname)) %>% 
  filter(grepl("BE_", name)) %>% 
  mutate(across(c(rowname, name), ~gsub("BAL_|BE_", "", .))) %>% 
  #Rename columns
  left_join(gene.key, by=c("name"="geneName")) %>% 
  mutate(name = paste0(hgnc_symbol, " (", name, ")")) %>% 
  select(-hgnc_symbol) %>% 
  pivot_wider() %>% 
  #rename rows
  left_join(gene.key, by=c("rowname"="geneName")) %>% 
  mutate(rowname = paste0(hgnc_symbol, " (", rowname, ")")) %>% 
  select(-hgnc_symbol) %>% 
  column_to_rownames() %>% 
  as.matrix()
```

#### Genes positively associated with eosinophils

There is additional notation if the gene was also significant for FeNO but this figure does not contain all FeNo significant genes.

```{r fig.height=8}
col.anno <- HeatmapAnnotation(
  `BE EOS FDR < 0.1` = filter(hm.mat.up, grepl("BE_", gene))$EOS,
  `BE FeNO FDR < 0.1` = filter(hm.mat.up, grepl("BE_", gene))$FeNO,
  col = list(`BE EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BE FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))
row.anno <- rowAnnotation(
  `BAL EOS FDR < 0.1` = filter(hm.mat.up, grepl("BAL_", gene))$EOS,
  `BAL FeNO FDR < 0.1` = filter(hm.mat.up, grepl("BAL_", gene))$FeNO,
  col = list(`BAL EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BAL FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))

set.seed(42)
hm.up <- Heatmap(corr.up.R, top_annotation = col.anno, left_annotation = row.anno,
        row_km = 3, column_km = 3, name = "Pearson",
        show_row_names = FALSE, use_raster = FALSE)
hm.up <- draw(hm.up)
```

#### Genes negatively associated with eosinophils

There is additional notation if the gene was also significant for FeNO but this figure does not contain all FeNo significant genes.

```{r fig.height=8}
col.anno2 <- HeatmapAnnotation(
  `BE EOS FDR < 0.1` = filter(hm.mat.dn, grepl("BE_", gene))$EOS,
  `BE FeNO FDR < 0.1` = filter(hm.mat.dn, grepl("BE_", gene))$FeNO,
  col = list(`BE EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BE FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))
row.anno2 <- rowAnnotation(
  `BAL EOS FDR < 0.1` = filter(hm.mat.dn, grepl("BAL_", gene))$EOS,
  `BAL FeNO FDR < 0.1` = filter(hm.mat.dn, grepl("BAL_", gene))$FeNO,
  col = list(`BAL EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BAL FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))

set.seed(42)
hm.dn <- Heatmap(corr.dn.R, top_annotation = col.anno2, left_annotation = row.anno2,
        row_km = 2, column_km = 2, name = "Pearson",
        show_row_names = FALSE, use_raster = FALSE)
hm.dn <- draw(hm.dn)
```

## FeNO

Allergen challenge (FDR < 0.3) and FeNO significant (FDR 0.1) genes in BAL vs BE. As expected, genes can be grouped into those that positively associate with FeNO in both BAL and BE (positive correlation) and those that have opposite directionality in BAL and BE (negative correlation).

```{r}
#Combine 
temp2 <- dat.BAL.rename %>% 
  filter(gene %in% c(BAL_feno_0.1_up, BAL_feno_0.1_dn)) %>% 
  mutate(EOS = ifelse(gene %in% c(BAL_eos_0.1_up, BAL_eos_0.1_dn), "FDR < 0.1", "NS"),
         FeNO = ifelse(gene %in% c(BAL_feno_0.1_up, BAL_feno_0.1_dn), "FDR < 0.1", "NS"))

dat.visit.all.rename2 <- dat.BE.rename %>% 
  filter(gene %in% c(BE_feno_0.1_up, BE_feno_0.1_dn)) %>% 
  mutate(EOS = ifelse(gene %in% c(BE_eos_0.1_up, BE_eos_0.1_dn), "FDR < 0.1", "NS"),
         FeNO = ifelse(gene %in% c(BE_feno_0.1_up, BE_feno_0.1_dn), "FDR < 0.1", "NS")) %>% 
  bind_rows(temp2) %>% 
  arrange(donorID) %>% 
  pivot_wider(names_from = donorID) 
```

```{r}
hm.mat.up2 <- dat.visit.all.rename2 %>% 
  select_if(~all(!is.na(.))) %>%
  filter(gene %in% c(BAL_feno_0.1_up,BE_feno_0.1_up))

hm.mat.up.mat2 <- hm.mat.up2 %>% 
  select(-EOS, -FeNO) %>% 
  column_to_rownames("gene")

hm.mat.dn2 <- dat.visit.all.rename2 %>% 
  select_if(~all(!is.na(.))) %>%
  filter(gene %in% c(BAL_feno_0.1_dn,BE_feno_0.1_dn)) 

hm.mat.dn.mat2 <- hm.mat.dn2 %>% 
  select(-EOS, -FeNO) %>% 
  column_to_rownames("gene")
```

```{r}
corr.up2 <- rcorr(t(hm.mat.up.mat2), type = "pearson")
corr.dn2 <- rcorr(t(hm.mat.dn.mat2), type = "pearson")

corr.up.R2 <- as.data.frame(corr.up2$r) %>% 
  rownames_to_column() %>% 
  pivot_longer(-rowname) %>% 
  #filter BAL vs BE
  filter(grepl("BAL_", rowname)) %>% 
  filter(grepl("BE_", name)) %>% 
  mutate(across(c(rowname, name), ~gsub("BAL_|BE_", "", .))) %>% 
  #Rename columns
  left_join(gene.key, by=c("name"="geneName")) %>% 
  mutate(name = paste0(hgnc_symbol, " (", name, ")")) %>% 
  select(-hgnc_symbol) %>% 
  pivot_wider() %>% 
  #rename rows
  left_join(gene.key, by=c("rowname"="geneName")) %>% 
  mutate(rowname = paste0(hgnc_symbol, " (", rowname, ")")) %>% 
  select(-hgnc_symbol) %>% 
  column_to_rownames() %>% 
  as.matrix()

corr.dn.R2 <- as.data.frame(corr.dn2$r) %>% 
  rownames_to_column() %>% 
  pivot_longer(-rowname) %>% 
  #filter BAL vs BE
  filter(grepl("BAL_", rowname)) %>% 
  filter(grepl("BE_", name)) %>% 
  mutate(across(c(rowname, name), ~gsub("BAL_|BE_", "", .))) %>% 
  #Rename columns
  left_join(gene.key, by=c("name"="geneName")) %>% 
  mutate(name = paste0(hgnc_symbol, " (", name, ")")) %>% 
  select(-hgnc_symbol) %>% 
  pivot_wider() %>% 
  #rename rows
  left_join(gene.key, by=c("rowname"="geneName")) %>% 
  mutate(rowname = paste0(hgnc_symbol, " (", rowname, ")")) %>% 
  select(-hgnc_symbol) %>% 
  column_to_rownames() %>% 
  as.matrix()
```

#### Genes positively associated with FeNO

There is additional notation if the gene was also significant for eosinophils but this figure does not contain all eosinophil significant genes.

```{r fig.height=8}
col.anno3 <- HeatmapAnnotation(
  `BE EOS FDR < 0.1` = filter(hm.mat.up2, grepl("BE_", gene))$EOS,
  `BE FeNO FDR < 0.1` = filter(hm.mat.up2, grepl("BE_", gene))$FeNO,
  col = list(`BE EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BE FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))
row.anno3 <- rowAnnotation(
  `BAL EOS FDR < 0.1` = filter(hm.mat.up2, grepl("BAL_", gene))$EOS,
  `BAL FeNO FDR < 0.1` = filter(hm.mat.up2, grepl("BAL_", gene))$FeNO,
  col = list(`BAL EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BAL FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))

set.seed(42)
hm.up2 <- Heatmap(corr.up.R2, top_annotation = col.anno3, left_annotation = row.anno3,
        row_km = 2, column_km = 2, name = "Pearson",
        show_row_names = FALSE, use_raster = FALSE)
hm.up2 <- draw(hm.up2)
```

#### Genes negatively associated with FeNO

There is additional notation if the gene was also significant for eosinophils but this figure does not contain all eosinophil significant genes.

```{r fig.height=8}
col.anno4 <- HeatmapAnnotation(
  `BE EOS FDR < 0.1` = filter(hm.mat.dn2, grepl("BE_", gene))$EOS,
  `BE FeNO FDR < 0.1` = filter(hm.mat.dn2, grepl("BE_", gene))$FeNO,
  col = list(`BE EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BE FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))
row.anno4 <- rowAnnotation(
  `BAL EOS FDR < 0.1` = filter(hm.mat.dn2, grepl("BAL_", gene))$EOS,
  `BAL FeNO FDR < 0.1` = filter(hm.mat.dn2, grepl("BAL_", gene))$FeNO,
  col = list(`BAL EOS FDR < 0.1` = c("FDR < 0.1" = "#E1BE6A", "NS" = "white"),
             `BAL FeNO FDR < 0.1` = c("FDR < 0.1" = "#40B0A6", "NS" = "white")), 
        show_legend = c(FALSE,FALSE))

set.seed(42)
hm.dn2 <- Heatmap(corr.dn.R2, top_annotation = col.anno4, left_annotation = row.anno4,
        row_km = 2, column_km = 2, name = "Pearson",
        show_row_names = FALSE, use_raster = FALSE)
hm.dn2 <- draw(hm.dn2)
```

### Heatmap results files

See `correlation_heatmap_clusters.csv` for gene membership in each cluster. The two heatmaps are designated as `eos_up`, `eos_down`, `feno_up`, or `feno_down`.

```{r}
source("get_hm_cluster.R")
cluster.row.eos.up <- get_hm_cluster(dat = corr.up.R, hm = hm.up, dimension = "row") %>% 
  mutate(cluster = paste("BAL_row",cluster,sep="_"),
         heatmap = "eos_up")
cluster.col.eos.up <- get_hm_cluster(dat = corr.up.R, hm = hm.up, dimension = "col") %>% 
  mutate(cluster = paste("BE_column",cluster,sep="_"),
         heatmap = "eos_up")
cluster.row.eos.dn <- get_hm_cluster(dat = corr.dn.R, hm = hm.dn, dimension = "row") %>% 
  mutate(cluster = paste("BAL_row",cluster,sep="_"),
         heatmap = "eos_down")
cluster.col.eos.dn <- get_hm_cluster(dat = corr.dn.R, hm = hm.dn, dimension = "col") %>% 
  mutate(cluster = paste("BE_column",cluster,sep="_"),
         heatmap = "eos_down")
####

cluster.row.feno.up <- get_hm_cluster(dat = corr.up.R2, hm = hm.up2, dimension = "row") %>% 
  mutate(cluster = paste("BAL_row",cluster,sep="_"),
         heatmap = "feno_up")
cluster.col.feno.up <- get_hm_cluster(dat = corr.up.R2, hm = hm.up2, dimension = "col") %>% 
  mutate(cluster = paste("BE_column",cluster,sep="_"),
         heatmap = "feno_up")
cluster.row.feno.dn <- get_hm_cluster(dat = corr.dn.R2, hm = hm.dn2, dimension = "row") %>% 
  mutate(cluster = paste("BAL_row",cluster,sep="_"),
         heatmap = "feno_down")
cluster.col.feno.dn <- get_hm_cluster(dat = corr.dn.R2, hm = hm.dn2, dimension = "col") %>% 
  mutate(cluster = paste("BE_column",cluster,sep="_"),
         heatmap = "feno_down")
```

```{r}
#Save
cluster.row <- bind_rows(cluster.row.eos.up, cluster.row.eos.dn,
                         cluster.row.feno.up,cluster.row.feno.dn) %>% 
  rename(gene=row, order_within_cluster=row_within_cluster)

cluster.col <- bind_rows(cluster.col.eos.up, cluster.col.eos.dn, 
                         cluster.col.feno.up, cluster.col.feno.dn) %>% 
  rename(gene=col, order_within_cluster=col_within_cluster)

bind_rows(cluster.row, cluster.col) %>% 
  select(heatmap, cluster, order_within_cluster, gene) %>% 
  arrange(heatmap, cluster) %>% 
  separate(gene, into = c("gene","hgnc_symbol")) %>% 
  write_csv("results/correlation_heatmap_clusters.csv")
```

# Gene plots

All genes significant for eosinophils and/or FeNO can be found in the zipped folder `figs/DEG/`

```{r eval=FALSE}
#signif genes to plot
genes.OI <- bind_rows(lm_eos_BAL$lm, lm_eos_BE$lm, lm_feno_BAL$lm, lm_feno_BE$lm) %>% 
  filter(variable != "(Intercept)" & FDR < 0.1) %>% 
  pull(gene) %>% unique()

#combine data
dat.BAL.all <- as.data.frame(dat.BAL$E) %>% 
  rownames_to_column("geneName") %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  left_join(dat.BAL$targets) %>% 
  left_join(dat.BAL$genes) %>% 
  mutate(gene_full_name = paste0(hgnc_symbol, " (", geneName, ")"))

dat.BE.all <- as.data.frame(dat.BE$E) %>% 
  rownames_to_column("geneName") %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  left_join(dat.BE$targets) %>% 
  left_join(dat.BE$genes) %>% 
  mutate(gene_full_name = paste0(hgnc_symbol, " (", geneName, ")"))

dat.all <- bind_rows(dat.BAL.all,dat.BE.all)

#Plot genes OI
for(i in 1:length(genes.OI)){
  print(paste(i, "of", length(genes.OI)))
  gene.temp <- genes.OI[i]

  #fdr
  fdr.bal.temp <- bind_rows(lm_visit_BAL$lm, lm_eos_BAL$lm, lm_feno_BAL$lm,) %>% 
    filter(variable != "(Intercept)") %>%
    filter(gene == gene.temp) %>% 
    mutate(Sample = "BAL") %>% 
    select(Sample, gene, variable, pval, FDR) 
  
  fdr.be.temp <- bind_rows(lm_visit_BE$lm, lm_eos_BE$lm, lm_feno_BE$lm) %>% 
    filter(variable != "(Intercept)") %>%
    filter(gene == gene.temp) %>% 
    mutate(Sample = "BE") %>% 
    select(Sample, gene, variable, pval, FDR)
  
  fdr.temp <- bind_rows(fdr.bal.temp, fdr.be.temp) %>% 
    mutate(across(c(pval, FDR), ~signif(., digits = 2))) %>% 
    select(gene, everything())

  dat.temp <- dat.all %>% 
    filter(geneName == gene.temp)
  
  #visit 
  p1 <- dat.temp %>% 
    ggplot(aes(x = visit, y = value)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2, height = 0) +
    facet_wrap(~group, scales = "free") +
    theme_classic() +
    labs(title = "Gene expression with allergen challenge",
         x = "Visit", y = "Log2 normalized expression")

  #EOS 
  p2 <- dat.temp %>% 
    filter(visit == "Post") %>% 
    ggplot(aes(x = EOS.pct, y = value)) +
    geom_point() +
    geom_smooth(method = "lm", formula = 'y ~ x', se = FALSE, color ="grey70") +
    facet_wrap(~group, scales = "free") +
    theme_classic() +
    labs(title = "Post-challenge gene expression and eosinophils",
         x = "Eosinophil (%)", y = "Log2 normalized expression")
  
  #FeNO 
  p3 <- dat.temp %>% 
    filter(visit == "Post") %>% 
    ggplot(aes(x = FeNO.PreBro_V5, y = value)) +
    geom_point() +
    geom_smooth(method = "lm", formula = 'y ~ x', se = FALSE, color ="grey70") +
    facet_wrap(~group, scales = "free") +
    theme_classic() +
    labs(title = "Post-challenge gene expression and FeNO",
         x = "FeNO", y = "Log2 normalized expression")
  
  p.all <- p1 + gridExtra::tableGrob(fdr.temp) + p2 + p3 + 
    plot_annotation(title = unique(dat.temp$gene_full_name))
  
  ggsave(paste0("figs/DEG/",
                gsub("\\(|\\)","",gsub(" ","_",unique(dat.temp$gene_full_name))),
                ".png"), p.all, width = 10, height = 8)
}
```

# R session

```{r}
sessionInfo()
```

***